docs(api): add profile picture url in attribute

haritianaadriano opened the pull request.

haritianaadriano committed:
docs(api): add profile picture url in attribute

docs(api): add description to profile picture attribute

docs(api): PUT /users/user_id/profile to upload or update user profile picture

docs(api): fix path parameters in user profile picture

docs(api): add manager and teacher endpoint for user profile picture

refactor: resources of Users profile picture endpoint

docs: refactor attribute description

docs: refactor response after uploading profile picture

docs(api): refactor last path to raw to insert profile picture

docs: refactor payload to separate creation and return

Mahefaa commented on doc/api.yml:
```diff
@@ -1832,6 +1832,8 @@ components:
           type: string
         nic:
           type: string
+        profile_picture:
+          type: string
```
unclear,
What kind of data will be stored here ? a link ? a filename ?

haritianaadriano replied:
I add some description to describe the attribute

Mahefaa commented on doc/api.yml:
```diff
@@ -1834,6 +1834,7 @@ components:
           type: string
         profile_picture:
           type: string
+          description: Profile picture of User stored in S3
```
this is not enough.

describe the format of the profile_picture (PNG ? JPEG?)
Does it contain the folderName ?
Are there restrictions concerning the way you name your file ?
Is it insertable ?

Mahefaa commented on doc/api.yml:
```diff
@@ -1687,6 +1687,117 @@ paths:
+              format: binary
+      responses:
+        '200':
+          description: Student picture uploaded
```
Student with updated profile picture suits better

Mahefaa commented on doc/api.yml:
```diff
@@ -1687,6 +1687,117 @@ paths:
+      tags:
+        - Users
+      summary: Upload teacher profile picture
+      operationId: uploadTeacherProfile
```
uploadTeacherProfilePicture !

Mahefaa commented on doc/api.yml:
```diff
@@ -1687,6 +1687,117 @@ paths:
+          required: true
+          schema:
+            type: string
+          description: Teacher id to upload profile picture
```
owner of the profile picture is simpler but i do not think you need a description at all.

Mahefaa commented on doc/api.yml:
why do you always return UserIdentifier in your responses ?

haritianaadriano replied:
i change to the corresponding payload as Teacher, Student or Manager model

Mahefaa commented on doc/api.yml:
```diff
@@ -1687,6 +1687,114 @@ paths:
           $ref: '#/components/responses/429'
         '500':
           $ref: '#/components/responses/500'
+  '/teachers/{teacher_id}/picture/upload':
```
use /raw instead of upload and we'll be good

haritianaadriano committed:
test: refactor test with the new payload during creation

fix: application build failed

test: refactor some test fo assertion

fix: test failed after refactoring payload

haritianaadriano committed:
fix: test failed after some refactor in payload

Mahefaa commented on doc/api.yml:
```diff
@@ -1915,19 +2027,35 @@ components:
+    CreateStudent:
+      allOf:
+        - $ref: '#/components/schemas/User'
+        - type: object
+          properties:
+            specialization_field:
+              $ref: '#/components/schemas/SpecializationField'
```
why is this in the allOf?

Mahefaa commented on doc/api.yml:
```diff
@@ -1915,19 +2027,35 @@ components:
           properties:
             specialization_field:
               $ref: '#/components/schemas/SpecializationField'
+        - $ref: '#/components/schemas/ProfilePictureUrl'
```
profilePictureUrl is too simple to be an Object

Mahefaa commented on src/main/java/school/hei/haapi/endpoint/rest/controller/ManagerController.java:
```diff
@@ -36,7 +37,7 @@ public Manager getManagerById(@PathVariable String id) {
   @PutMapping("/managers/{id}")
   public Manager updateManager(
-      @PathVariable(name = "id") String managerId, @RequestBody Manager toUpdate) {
+      @PathVariable(name = "id") String managerId, @RequestBody CreateManager toUpdate) {
```
CreateManager does not seem correct if it also serves for update

Mahefaa commented on doc/api.yml:
```diff
@@ -1915,19 +2027,35 @@ components:
+      allOf:
+        - $ref: '#/components/schemas/User'
     Student:
       allOf:
         - $ref: '#/components/schemas/User'
         - type: object
           properties:
             specialization_field:
               $ref: '#/components/schemas/SpecializationField'
```
isn't this CreateStudent ?

Mahefaa commented on doc/api.yml:
try something along these lines : 

to create an User, you have to insert all the properties of CrupdateUser.
ex : to create or update a student you need all properties of CrupdateUser + the specialization_field.
You are allowed to create new components such as CrupdateStudent, CrupdateManager, CrupdateTeacher.

And then you create 3 different components who are : Student, Teacher, Manager.

Each one of these 3 components will inherit the properties of their Crupdate version.

And then, you will add profile_picture attribute (as a String) .

haritianaadriano replied:
ok, i finished to refactor this part

haritianaadriano committed:
refactor: some model in docs payload

temp

Mahefaa commented on doc/api.yml:
```diff
@@ -1915,19 +2023,41 @@ components:
         - $ref: '#/components/schemas/User'
+        - properties:
+            profile_picture:
+              type: string
```
why is it in the allOf ?

haritianaadriano committed:
docs: refactor properties emplacement for user

temp

Mahefaa commented on doc/api.yml:
```diff
@@ -1915,19 +2023,40 @@ components:
+    CrupdateStudent:
       allOf:
         - $ref: '#/components/schemas/User'
         - type: object
           properties:
             specialization_field:
               $ref: '#/components/schemas/SpecializationField'
```
move this outside allOf scope

Mahefaa commented on doc/api.yml:
```diff
@@ -1915,19 +2023,40 @@ components:
+    CrupdateManager:
+      allOf:
+        - $ref: '#/components/schemas/User'
+    Student:
+      allOf:
+        - $ref: '#/components/schemas/User'
+        - type: object
+      properties:
+        specialization_field:
+          $ref: '#/components/schemas/SpecializationField'
```
this is CrupdateUser isn't it ?
use allOf CrupdateUser then

Mahefaa commented on src/test/java/school/hei/haapi/integration/TeacherIT.java:
```diff
@@ -168,24 +170,27 @@ void manager_write_update_rollback_on_event_error() throws ApiException {
     ApiClient manager1Client = anApiClient(MANAGER1_TOKEN);
-    Teacher toCreate = someCreatableTeacher();
+    CrupdateTeacher toCreate = someCreatableTeacher();
+    Teacher expected = expectedSomeTeacherCreated();
```
if not reused in other classes then put it in this class as private instead of TestUtils

Mahefaa replied:
also, rename to expectedCreatedTeacher

Mahefaa reviewed (changes requested):
last changes, keep going :))

Mahefaa reviewed (approved):
LGTM
